### ISSUE TITLE
Using Rclone mount with iconv module fails to see or create some filenames

### ISSUE THREAD

**Issue by FurloSK** [rclone/rclone#8088]
#### Background
After issues #7072 and #7620 have been solved and merged into Rclone 1.67, Unicode not fully works under macOS. Yay!

I wanted to utilize `iconv` module to make it so that any files created on my Mac will be transferred to remote in NFC normalisation form (that is, in composed form). This is easily done with the following switch during the `mount` command:

    -o modules=iconv,from_code=UTF-8,to_code=UTF-8-MAC

I have checked that it works in Google Drive. When I create a folder called "čačaný" on my Mac using the above `iconv` switch, it will appear in Google Drive (accessed via web browser) in NFC form. Without the `iconv` module switch, it will appear in NFD form.

#### The problem
**_Using the above `iconv` switch causes files containing in their filename some special characters_** (e.g. 📖, 📕) **_to be completely ignored by Rclone._**

These files are missing in Rclone folder listings, but they do not even appear in Rclone log when I enter the folder containing them for the first time (not even with `--log-level DEBUG`). It is as if they never existed. Trying to create a folder containing such characters (e.g. "test📖") causes the following error in Finder:

**`The operation can’t be completed.`**
`An unexpected error occurred (error code 100092).`

In Path Finder, trying to do the same gets you:

**`New folder with name "test📖" cannot be created.`**
`The operation couldn’t be completed. Illegal byte sequence`
`Illegal byte sequence`

But again, there is nothing in the Rclone log – as if the attempt to create the folder was never made in the first place.

What could be done with this? What is so special about these characters that causes Rclone to misbehave in this way? And how can I solve this? I suspect that it is an underlying problem (with iconv? or Fuse?), seeing that it never gets into Rclone log, but I do not have enough knowledge to trace that problem further.

Also, there seems to be a problem logged for Rsync project which might be relevant:
https://github.com/RsyncProject/rsync/issues/77

### What is your rclone version (output from `rclone version`)
rclone v1.67.0

### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
macOS Mojave (darwin 10.14.6, 64 bit)

### Which cloud storage system are you using? (e.g. Google Drive)
Google Drive

### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
    rclone mount MY_REMOTE: /path/to/mount/point -o modules=iconv,from_code=UTF-8,to_code=UTF-8-MAC --log-level DEBUG --vfs-cache-mode full

### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by erusc** [comment]
Now that there is `--name-transform`, I image that this can be done on rclone side.
I can add `--name-transform "all,nfc"` to ensure files copied to the remote is in NFC form.
```sh
rclone copy ./MY_DIR/ MY_REMOTE: --name-transform "all,nfc"
```

Unfortunately `--name-transform` does not work with `mount` command currently.
```sh
rclone mount MY_REMOTE: /path/to/mount/point --name-transform "all,nfc"
```

I hope that this can be done without libfuse because I mount remotes on a Linux server which works as a intermediate server accepting uploads from various clients including Apple devices.

@nielash, is it possible to make `--name-transform` work with mount?

**Comment by nielash** [comment]
> [@nielash](https://github.com/nielash), is it possible to make `--name-transform` work with mount?

Currently it is not designed to work with mount. It might be possible to add mount support, but there would be some conceptual problems to think through. 

For instance, consider a simple example of adding a suffix `_suf`. Is the idea that creating `file.txt` in the mount should actually copy the file as `file_suf.txt` to the underlying remote? If so, doesn't it then follow that it should be displayed as `file_suf.txt` when the remote is mounted? And then, if it is edited, does that create `file_suf_suf.txt` etc., infinitely?

These kinds of problems arise because `--name-transform` is designed for a one-way sync, whereas `mount` is bidirectional.

In the NFC / NFD scenario, it might help if you could walk me through what you're hoping will happen. For example, a mac user creates a file with an NFD name in a mount... you want that to be copied to the underlying remote as the NFC equivalent, correct? And then in the other direction, if the remote has an NFC file, what should happen in the mount on the mac side? Attempt to display it as NFC? Preemptively convert it to NFD?

Would it solve your problem to just run:

```
rclone convmv MY_REMOTE: --name-transform "all,nfc"
```

on your remote every so often, to convert any NFD names to your preferred NFC?

Also, regarding the OP:

> #### The problem
> **_Using the above `iconv` switch causes files containing in their filename some special characters_** (e.g. 📖, 📕) **_to be completely ignored by Rclone._**

I'm not quite following why you need to use the `iconv` switch at all, when #7620 was intended to eliminate the need for `-o modules=iconv`. Are you mounting on mac with `--no-unicode-normalization=false` (the default)? If so, what about that is not working for you, that you hope `iconv` will solve?

There may be a better solution here, but first I think I need to better understand the problem 🙂 

CC @kapitainsky in case you have thoughts -- I know you are an expert on this issue!

**Comment by erusc** [comment]
> Would it solve your problem to just run:
> 
> ```
> rclone convmv MY_REMOTE: --name-transform "all,nfc"
> ```
> 
> on your remote every so often, to convert any NFD names to your preferred NFC?

Yes, but I wish it to happen automatically whenever a file is uploaded.
BTW, this command does not work well with OneDrive.
OneDrive allows both NFD and NFC on their server, but when I try to convert NFD to NFC, it consider them to be the same path and thus cannot be renamed.
I had to move it elsewhere and back again.

Dropbox, for example, handles this on their servers.
NFD files uploaded from any client (even third-party) are always converted to NFC.

**Comment by nielash** [comment]
> Yes, but I wish it to happen automatically whenever a file is uploaded.

Got it. I am wondering now if perhaps the best way to implement this would be with a new [`--backend-encoding`](https://rclone.org/overview/#encoding) parameter to (optionally) force NFC (or NFD) at the backend level. Thoughts on that idea?

One thing that's tricky is that the encoder usually assumes the encoding will be losslessly invertible. That is not the case with NFC / NFD (because multiple inputs can have the same output). I'm not sure if that will cause problems -- we'd need to test that. But it's possible that `InvalidUtf8` already works that way.

> BTW, this command does not work well with OneDrive.
> OneDrive allows both NFD and NFC on their server, but when I try to convert NFD to NFC, it consider them to be the same path and thus cannot be renamed.
> I had to move it elsewhere and back again.

Thank you for reporting this. Can you make a new issue about this and tag me? I think it should be easy to fix.

**Comment by erusc** [comment]
> Got it. I am wondering now if perhaps the best way to implement this would be with a new [`--backend-encoding`](https://rclone.org/overview/#encoding) parameter to (optionally) force NFC (or NFD) at the backend level. Thoughts on that idea?

I have never used the `encoding` parameter, but it does sound great if it can be set in the config file.
It eliminates the need to modify existing scripts and add an extra flag to future commands.

**Comment by nielash** [comment]
> I have never used the `encoding` parameter, but it does sound great if it can be set in the config file.
> It eliminates the need to modify existing scripts and add an extra flag to future commands.

I've created a proof-of-concept here: https://github.com/nielash/rclone/commit/a8bb97c1cc1b74d0a2f11d9e2f90383247f48436

Would you be willing to test it out and report back?

**Comment by erusc** [comment]
It works! That is exactly what I need.
It it possible to make it work with `crypt`?
I have tried this with no luck.
```
[OneDrive2]
type = onedrive
...
encoding = ForceNFC

[OneDrive2Crypt]
type = crypt
remote = OneDrive2:
filename_encryption = obfuscate
...
encoding = ForceNFC
```

It also works with mount but there is a potential integration issue because it only enforces the encoding on the remote.
This may be fine on macOS but not for my use case (a Linux server accepting uploads from macOS/iOS).
The local path is still in NFD when accessed directly or via other applications until the [VFS directory cache](https://rclone.org/commands/rclone_mount/#vfs-directory-cache) expires (as controlled by `--dir-cache-time` and `--poll-interval`).
This also applies to a Dropbox mount with a stable version of rclone currently.

**Comment by nielash** [comment]
> It works! That is exactly what I need.

Great!

> It it possible to make it work with `crypt`?
> I have tried this with no luck.

I think your issue is the following:
```
filename_encryption = obfuscate
```

With the default of `standard`, crypt would use the base32 alphabet for encrypted filenames, which is limited to simple ASCII characters that are identical between NFD and NFC. But `obfuscate` just rotates the existing filename characters. `obfuscate` is also not truly encryption and is not very secure. Is there a reason you need to use it?

I don't think we can simply normalize an obfuscated filename, as `crypt` would likely have trouble decoding it.

Note that if you want to change to `standard`, but you have already encrypted files with `obfuscate`, it is best to make a new `crypt` remote and copy the files from the old to the new. Otherwise `crypt` will not be able to decode the existing files.



> It also works with mount but there is a potential integration issue because it only enforces the encoding on the remote.
> This may be fine on macOS but not for my use case (a Linux server accepting uploads from macOS/iOS).
> The local path is still in NFD when accessed directly or via other applications until the [VFS directory cache](https://rclone.org/commands/rclone_mount/#vfs-directory-cache) expires (as controlled by `--dir-cache-time` and `--poll-interval`).

Hmm. If you add `--local-encoding "ForceNFC"` does that change anything? If not, this probably needs to be handled in the VFS layer.

**Comment by erusc** [comment]
> With the default of `standard`, crypt would use the base32 alphabet for encrypted filenames, which is limited to simple ASCII characters that are identical between NFD and NFC. But `obfuscate` just rotates the existing filename characters. `obfuscate` is also not truly encryption and is not very secure. Is there a reason you need to use it?

I had to use obfuscate because the length of encrypted filename exceeds OneDrive's limit.
https://support.microsoft.com/en-us/office/restrictions-and-limitations-in-onedrive-and-sharepoint-64883a5d-228e-48f5-b3d2-eb39e07630fa#filenamepathlengths
> The length of the OneDrive root folder (e.g. C:\users\meganb\OneDrive - Contoso) + the relative path of the file (up to 400 chars) cannot be more than 520 characters.

I have also tried with Google Drive, which allows longer filenames and works fine with encrypted filenames, and it is still NFD on the remote when viewing with `./rclone-encoder-nfc ncdu GDriveCrypt:`.
```
[GDrive]
type = drive
client_id = ***
client_secret = ***
token = ***
encoding = ForceNFC

[GDriveCrypt]
type = crypt
remote = GDrive:
password = ***
password2 = ***
encoding = ForceNFC
```
It is probably meaningless to put `encoding = ForceNFC` under the base remote because encrypted filenames are alphanumeric.
I could not find `--crypt-encoding` in https://rclone.org/crypt/ or https://rclone.org/flags/ so I did not expect it to work when I tried.

---

> Hmm. If you add --local-encoding "ForceNFC" does that change anything? If not, this probably needs to be handled in the VFS layer.

No, it does not change anything.
```sh
./rclone-encoder-nfc mount --local-encoding "ForceNFC" OneDrive: ./OneDrive
```
I suppose it only works for `type = local`.
```
[local]
type = local
encoding = ForceNFC
```

**Comment by nielash** [comment]
> I could not find `--crypt-encoding` in https://rclone.org/crypt/ or https://rclone.org/flags/ so I did not expect it to work when I tried.

Right, `crypt` definitely does not support `encoding` at the moment -- my point was just that usually this wouldn't matter, because the base32 alphabet is limited to characters that are the same in both NFC and NFD. But now I think perhaps I misunderstood and thought you were talking about the _encrypted_ filename, when actually you were talking about the _decrypted_ filename.

Perhaps `crypt` should add support for `encoding`? I think this doesn't come up more often because `encoding` is usually used to get around filesystem limitations, not to enforce stricter restrictions beyond what the filesystem requires. This isn't usually necessary with `crypt`, because of the very limited base32 alphabet. In fact, `crypt` itself is sometimes used as a way to encode filenames that wouldn't normally be allowed on the underlying remote. But off the top of my head, I can't think of a reason why `crypt` _couldn't_ support the multiencoder.

Another idea: perhaps there should be a new `encoder` wrapping backend that just encodes/decodes filenames? Then you could just wrap `crypt` (or anything else) with this, to make sure the names are NFC before `crypt` ever sees them.

> I had to use obfuscate because the length of encrypted filename exceeds OneDrive's limit.
> https://support.microsoft.com/en-us/office/restrictions-and-limitations-in-onedrive-and-sharepoint-64883a5d-228e-48f5-b3d2-eb39e07630fa#filenamepathlengths

OneDrive is indeed quite limiting, particularly in counting the entire path rather than just the leaf. There are many OneDrive users who use `crypt`, although I think many of them just make sure to use short filenames.

One possible idea you might consider is: 
```
filename_encoding = base32768
```
which results in names significantly shorter than `base64` or `base32`. Be aware, though, that there is a theoretical risk of a case-insensitivity collision -- albeit one that is statistically unlikely to occur in the real world. Interesting thread about that here: https://github.com/rclone/rclone/issues/6803

Also worth mentioning that there is a new `--name-transform truncate_bytes_keep_extension=N` mode available in beta. https://github.com/rclone/rclone/commit/fe62a2bb4ea8f62a1567c7883c0bba81827656eb Quite useful for automatically truncating long names.

> I suppose it only works for `type = local`.

That makes sense. It is probably something that would need to be handled in the VFS layer then. I can see if there's a minimally-invasive way of doing that.

**Comment by FurloSK** [comment]
Hi guys, sorry for a late reply!

> > > In the NFC / NFD scenario, it might help if you could walk me through what you're hoping will happen. […] Would it solve your problem to just run:
> > > ```
> > > rclone convmv MY_REMOTE: --name-transform "all,nfc"
> > > ```
> > > on your remote every so often, to convert any NFD names to your preferred NFC?
> > 
> > Yes, but I wish it to happen automatically whenever a file is uploaded.
> 
> Got it. I am wondering now if perhaps the best way to implement this would be with a new [`--backend-encoding`](https://rclone.org/overview/#encoding) parameter to (optionally) force NFC (or NFD) at the backend level. Thoughts on that idea?

Indeed, this is exactly what I want to achieve. **I am on Mac and I want all the filenames to be created/renamed in NFC form** for [these reasons](https://www.furlo.sk/blog/odborny/2024-09-22-unicode_nfc_normalisation_for_rclone_on_macos#problems_raised_by_the_difference_in_normalisation_forms).

The process which I currently use to achieve this (and which I described at length in [this blog post](https://www.furlo.sk/blog/odborny/2024-09-22-unicode_nfc_normalisation_for_rclone_on_macos)) is to use iconv during rclone mount with the option `-o modules=iconv,from_code=UTF-8,to_code=UTF-8-MAC`.

This forces all files output by rclone to cloud to be in NFC, but also causes the mentioned problem of missing filenames. I suspect MacOS crippled version of iconv is to blame here (again, see the linked blog post above for details). However, having just a neat switch directly in `rclone mount` command to achieve this without adding iconv to the process would simplify this a lot!

**Comment by nielash** [comment]
Thank you @FurloSK, and thank you also for that very helpful blog post! 

I'm curious if you've tried it _without_ the `-o modules=iconv,from_code=UTF-8,to_code=UTF-8-MAC` option in the past year or so (after https://github.com/rclone/rclone/pull/7620 was merged)? It seems to me that should at least solve the portion of problems caused by Apple's buggy and incomplete `iconv` implementation, such as the "missing filenames" issue. This is because it relies on the https://pkg.go.dev/golang.org/x/text/unicode/norm library for conversion instead of Apple's version. 

Secondly, you might consider trying the beta I mentioned here: https://github.com/rclone/rclone/issues/8088#issuecomment-3003140547 This would handle the actual renaming on the remote (as opposed to just normalizing when comparing between remotes). https://github.com/rclone/rclone/issues/8088#issuecomment-3004676267 said it worked well (subject to the caveat about the `--dir-cache-time`).

I'm still interested in the idea of forcing a rename in the mount itself, but that would be a more complicated change. And would be potentially prone to odd behavior (if, for example, it gets renamed while the file is open and being edited, etc.)

**Comment by FurloSK** [comment]
@nielash :
> I'm curious if you've tried it _without_ the `-o modules=iconv,from_code=UTF-8,to_code=UTF-8-MAC` option in the past year or so (after [#7620](https://github.com/rclone/rclone/pull/7620) was merged)?

Yes, I actually tested rclone without the flag when I was updating to version 1.69.3 to see whether #7620 did not solve my problem. While it solved displaying of characters (that is, everything was listed, even filenames with emoji like 📖), when I wanted to create files only in NFC, I had to revert to the infamous `-o modules=iconv,from_code=UTF-8,to_code=UTF-8-MAC` which again caused missing files (and that's how that lengthy blog post of mine got created).

It would be great if `encoding = ForceNFC` could be used for this purpose, instead of that nightmare-like process I use. **I am definitely willing to test it, have you created a test build?** (I could not find it)

**Comment by nielash** [comment]
> It would be great if `encoding = ForceNFC` could be used for this purpose, instead of that nightmare-like process I use. **I am definitely willing to test it, have you created a test build?** (I could not find it)

@FurloSK Test builds are uploading now, you should find them here in about 20 minutes: 
https://beta.rclone.org/branch/encoder-nfc/v1.71.0-beta.8896.24835344e.encoder-nfc/

Branch: https://github.com/rclone/rclone/tree/encoder-nfc
